src/wallet/wallet.cpp

```diff
@@ -2816,195 +2817,99 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransac
-                            nChangePosInOut = GetRandInt(txNew.vout.size()+1);
-                        }
-                        else if ((unsigned int)nChangePosInOut > txNew.vout.size())
-                        {
-                            strFailReason = _("Change index out of range");
-                            return false;
-                        }
-
-                        std::vector<CTxOut>::iterator position = txNew.vout.begin()+nChangePosInOut;
-                        txNew.vout.insert(position, newTxOut);
-                    }
-                } else {
-                    nChangePosInOut = -1;
-                }
-
-                // Dummy fill vin for maximum size estimation
-                //
-                for (const auto& coin : setCoins) {
-                    txNew.vin.push_back(CTxIn(coin.outpoint,CScript()));
-                }
-
-                nBytes = CalculateMaximumSignedTxSize(txNew, this);
-                if (nBytes < 0) {
-                    strFailReason = _("Signing transaction failed");
+                        strFailReason = _("Transaction amount too small");
                     return false;
                 }
+                txNew.vout.push_back(txout);
+            }
 
-                nFeeNeeded = GetMinimumFee(*this, nBytes, coin_control, ::mempool, ::feeEstimator, &feeCalc);
-                if (feeCalc.reason == FeeReason::FALLBACK && !m_allow_fallback_fee) {
-                    // eventually allow a fallback fee
-                    strFailReason = _("Fee estimation failed. Fallbackfee is disabled. Wait a few blocks or enable -fallbackfee.");
-                    return false;
-                }
+            // Choose coins to use
+            bool bnb_used;
+            CAmount nValueIn = 0;
+            setCoins.clear();
+            coin_selection_params.change_spend_size = CalculateMaximumSignedInputSize(change_prototype_txout, this);
+            coin_selection_params.effective_fee = nFeeRateNeeded;
+            if (!SelectCoins(vAvailableCoins, nValueToSelect, setCoins, nValueIn, coin_control, coin_selection_params, bnb_used))
+            {
+                strFailReason = _("Insufficient funds");
+                return false;
+            }
+
+            const CAmount nChange = nValueIn - nValueToSelect;
+            if (nChange > 0)
+            {
+                // Fill a vout to ourself
+                CTxOut newTxOut(nChange, scriptChange);
 
-                // If we made it here and we aren't even able to meet the relay fee on the next pass, give up
-                // because we must be at the maximum allowed fee.
-                if (nFeeNeeded < ::minRelayTxFee.GetFee(nBytes))
+                // Never create dust outputs; if we would, just
+                // add the dust to the fee.
+                // The nChange when BnB is used is always going to go to fees.
+                if (IsDust(newTxOut, discard_rate) || bnb_used)
```

Comment:
wouldn't the excess from selecting with bnb always be below the discard_rate as well?